New PSCi command: reload

noraesae opened the pull request:
Regarding #2713 

This PR update the list of commands in PSCi. The current `:reset` command is renamed to `:clear`, and a new command `:reload` is added. `:reload` is expected to work the same as `:reload` in GHCi.

paf31 commented on src/Language/PureScript/Interactive.hs:
```diff
@@ -217,12 +226,15 @@ handleImport
      Left errs -> printErrors errs
      Right _  -> put st
+   where
+   renew :: ImportedModule -> [ImportedModule] -> [ImportedModule]
```
What is this function for?

noraesae replied:
It's to prevent import duplication. Previously, when a module was imported twice, it was listed in the imported modules twice. Maybe the name renew is somewhat confusing?

paf31 replied:
> Previously, when a module was imported twice, it was listed in the imported modules twice.

Is that necessary? Maybe that's what we need to fix?

noraesae replied:
Yes, that's what I fixed with the patch cd13ee4234203e9faed27c488ca55b7edc8e9e49 included in this PR.

Before:

```
> import Data.Functor
> import Data.Functor
> import Data.Functor
> import Data.Functor
> :show import
import Data.Functor
import Data.Functor
import Data.Functor
import Data.Functor

>
```

After:

```
> import Data.Functor
> import Data.Functor
> import Data.Functor
> import Data.Functor
> :show import
import Data.Functor

>
```

paf31 replied:
I think we want to allow multiple imports with the same name though. Otherwise you can't do something like import something qualified and with selective imports at the same time.

noraesae replied:
Oh, right.

Then maybe we can only renew when it's really a duplication. How do you think about it?

noraesae replied:
I've updated the PR. A comment is below.

paf31 replied:
I don't understand why we need `renew` at all.

noraesae replied:
Does it mean that exactly same imports can be allowed and be stacked in the import list? I'm afraid I do not yet understand the reason to keep the duplications.

paf31 replied:
If the user adds duplicate imports, that's fine. When the user runs the `:r` command, we should keep all of those imports, but remove the bindings.

noraesae replied:
I understand users can add duplicate imports and PSCi works without any problem. However, according to my understanding, duplicate imports mean nothing, but they are all listed in `:show import`. I still think we do not need to keep the duplicates.

It's actually not about `:r`, but `import X` in PSCi. May I revert it from this PR and open a discussion?

paf31 replied:
> However, according to my understanding, duplicate imports mean nothing

No, they are useful:

```
> import Prelude ((+))
> import Prelude as P
```

This means that I want to use `+` globally but qualify everything else. If the user typed that, we should preserve both.

noraesae replied:
Yes, but my update was removing only the identical ones.

```
> import Prelude ((+))
> import Prelude as P
> :show import
import Prelude ((+))
import Prelude as P

> import Prelude
> import Prelude
> import Prelude ((+))
> import Prelude as P
> :show import
import Prelude
import Prelude ((+))
import Prelude as P
```

paf31 replied:
I think if we were going to do that, we should consolidate any imports under the same namespace. So for example

```
import Prelude ((+))
import Prelude (bind)
import Prelude as P ((*))
import Prelude as P (show)
```

should give

```
> :show import
import Prelude ((+), (*))
import Prelude as P (bind, show)
```

Without that, I'm not sure it's worth it to be honest.

noraesae replied:
Your suggestion looks better. I may upload an issue for the consolidation, and revert the `renew` from PR. Thanks again for your review 👍

paf31 commented:
Looks great! Could you please add a test to `tests/TestPsci.hs` as well?

noraesae commented:
Will add a test.

noraesae commented:
Updated the renew to not only check the name, but the `ImportedModule` tuple itself. By this change, multiple imports with different forms work as expected:

```
> import Prelude
> import Prelude ((+))
> import Prelude
> import Prelude
> import Prelude as P
> import Prelude
> import Prelude ((+))
> import Prelude
> :show import
import Prelude
import Prelude ((+))
import Prelude as P

>
```

Also, I've tried to add directive tests. To make it actually work like actual PSCi, I've added a build steps, but `stack test` seems like failing.

```
Cases: 30  Tried: 11  Errors: 11  Failures: 0Compiling Data.Maybe
MultipleErrors {runMultipleErrors = [ErrorMessage [ErrorInModule (ModuleName [ProperName {runProperName = "Data"},ProperName {runProperName = "Maybe"}]),PositionedError (SourceSpan {spanName = "tests/support/bower_components/purescript-maybe/src/Data/Maybe.purs", spanStart = SourcePos {sourcePosLine = 11, sourcePosColumn = 1}, spanEnd = SourcePos {sourcePosLine = 11, sourcePosColumn = 27}}),PositionedError (SourceSpan {spanName = "tests/support/bower_components/purescript-maybe/src/Data/Maybe.purs", spanStart = SourcePos {sourcePosLine = 11, sourcePosColumn = 17}, spanEnd = SourcePos {sourcePosLine = 11, sourcePosColumn = 26}})] (UnknownImport (ModuleName [ProperName {runProperName = "Data"},ProperName {runProperName = "Eq"}]) (TyClassName (ProperName {runProperName = "Eq1"})))]}
```

I guess it's because of recent addition of `Eq1` into `Prelude`. I've tried to fix this by modifying `bower.json` in the test support dir, but no success. Any suggestion?

noraesae commented:
There have been some errors about compiling `Eq1` and `Ord1` but they are fixed by `bower cache clean`.

noraesae commented:
Reverted duplication removal for PSCi imports.
